stan/math/prim/mat/prob/normal_id_glm_lpdf.hpp

```diff
@@ -20,138 +21,160 @@ namespace math {
-  {
-    scalar_seq_view<T_n> n_vec(n);
-    scalar_seq_view<T_scale> sigma_vec(sigma);
-    for (size_t n = 0; n < N; ++n) {
-      sigma_dbl[n] = value_of(sigma_vec[n]);
-      n_dbl[n] = value_of(n_vec[n]);
-    }
+  scalar_seq_view<T_scale> sigma_vec(sigma);
+  Array<T_partials_return, Dynamic, 1> inv_sigma(N, 1);
+  for (size_t m = 0; m < N; ++m) {
+    inv_sigma[m] = 1 / value_of(sigma_vec[m]);
   }
-  Array<T_partials_return, Dynamic, 1> inv_sigma = 1 / sigma_dbl;
   Matrix<T_partials_return, Dynamic, 1> beta_dbl(M, 1);
   {
     scalar_seq_view<T_beta> beta_vec(beta);
     for (size_t m = 0; m < M; ++m) {
       beta_dbl[m] = value_of(beta_vec[m]);
     }
   }
-  Matrix<T_partials_return, Dynamic, Dynamic> x_dbl = value_of(x);
 
   Array<T_partials_return, Dynamic, 1> mu_dbl
-      = (x_dbl * beta_dbl
-         + Matrix<double, Dynamic, 1>::Ones(N, 1) * value_of(alpha))
-            .array();
-  Array<T_partials_return, Dynamic, 1> n_minus_mu_over_sigma
-      = (n_dbl - mu_dbl) * inv_sigma;
-  Array<T_partials_return, Dynamic, 1> n_minus_mu_over_sigma_squared
-      = n_minus_mu_over_sigma.square();
+      = (value_of(x) * beta_dbl).array();
+  scalar_seq_view<T_alpha> alpha_vec(alpha);
+  for (size_t m = 0; m < N; ++m)
+    mu_dbl[m] += value_of(alpha_vec[m]);
+  Array<T_partials_return, Dynamic, 1> y_minus_mu_over_sigma
+      = (value_of(y).array() - mu_dbl) * inv_sigma;
+  Array<T_partials_return, Dynamic, 1> y_minus_mu_over_sigma_squared
+      = y_minus_mu_over_sigma.square();
 
   if (include_summand<propto>::value)
     logp += NEG_LOG_SQRT_TWO_PI * N;
-  if (include_summand<propto, T_scale>::value)
-    logp -= sigma_dbl.log().sum();
-  if (include_summand<propto, T_n, T_x, T_beta, T_alpha, T_scale>::value)
-    logp -= 0.5 * n_minus_mu_over_sigma_squared.sum();
+  if (include_summand<propto, T_scale>::value) {
+    for (size_t m = 0; m < N; ++m) {
+      logp -= log(value_of(sigma_vec[m]));
+    }
+  }
+  if (include_summand<propto, T_y, T_x, T_alpha, T_beta, T_scale>::value)
+    logp -= 0.5 * y_minus_mu_over_sigma_squared.sum();
 
   // Compute the necessary derivatives.
-  operands_and_partials<T_n, T_x, T_beta, T_alpha, T_scale> ops_partials(
-      n, x, beta, alpha, sigma);
-  if (!(is_constant_struct<T_n>::value && is_constant_struct<T_x>::value
+  operands_and_partials<T_y, T_x, T_alpha, T_beta, T_scale> ops_partials(
+      y, x, alpha, beta, sigma);
+  if (!(is_constant_struct<T_y>::value && is_constant_struct<T_x>::value
         && is_constant_struct<T_beta>::value
         && is_constant_struct<T_alpha>::value)) {
     Matrix<T_partials_return, Dynamic, 1> mu_derivative
-        = (inv_sigma * n_minus_mu_over_sigma).matrix();
-    if (!is_constant_struct<T_n>::value) {
-      ops_partials.edge1_.partials_ = -mu_derivative;
+        = (inv_sigma * y_minus_mu_over_sigma).matrix();
+    if (!is_constant_struct<T_y>::value) {
+      assign_to_matrix_or_broadcast_array(ops_partials.edge1_.partials_,
```

Comment:
Aww man I had one more comment that I musta messed up and not included last time. And it's the Really Big One :P. Would it make sense to, instead of defining assign_to_matrix_or_broadcast_array, to define some sort of function:

```
ops_partials.edge1_.set_partials(-mu_derivative);
```

Cause presumably the is-a-matrix or is-a-broadcast-array comes from what sort of edge you're assigning to. If there was a member function for each type of edge that just Did The Right Thing, wouldn't that be better than this weird assign_to_matrix_or_broadcast_array operator?

Reply:
Ah, that is a great suggestion. Don't know why I didn't do it that way the first time around. Thanks for being so thorough!